Docs: Add documentation for Pod Disruption Budgets

scholzj opened the pull request:
### Type of change

- Documentation

### Description

This PR delivers the documentation related to #1199 

### Checklist

- [x] Update documentation

tombentley commented on documentation/book/con-customizing-pod-disruption-budgets.adoc:
```diff
@@ -0,0 +1,47 @@
+By default, the Pod Disruption Budgets always allow up to one Pod to be unavailable at a time by setting the `maxUnavailable` value in `PodDisruptionBudget.spec` to `1`.
+You can change this default setting using the Pod Disruption Budget template.
+
+In the Pod Disruption Budget template you can - in addition to Labels and Annotations - customize the `maxUnavailable` field as well:
```
This sticks out at me as being a restatement of the previous sentence. I think we could remove this.

tombentley commented on documentation/book/con-customizing-pod-disruption-budgets.adoc:
```diff
@@ -0,0 +1,47 @@
+
+In the Pod Disruption Budget template you can - in addition to Labels and Annotations - customize the `maxUnavailable` field as well:
+
+[table,stripes=none]
```
I have several problems with this table. 

* It has a subset of the properties of the one in the API reference, but it's not clear that the properties in this table is not exhaustive.
* It duplicates the documentation of `maxUnavailable` using different words. This is a pattern we're already using elsewhere, but it means there's content which needs maintaining in two places, which is a maintenance headache in the long run.

If we're really going to stick with this pattern of having concepts duplicating the content of the API reference then I think we should do it in a more intelligent way. For example, we could generate a `fragment-PodDisruptionBudgetTemplate.adoc` file with this table in it. That would allow you to write useful verbiage around the configurable properties without needing to duplicate the tables.

tombentley commented on documentation/book/con-customizing-pod-disruption-budgets.adoc:
```diff
@@ -0,0 +1,47 @@
+# ...
+template:
+    podDisruptionBudget:
+        metadata:
```
In reviewing this I looked again at the `PodDistruptionBugetTemplate` and I think the `@Description` on the `PodDistruptionBugetTemplate.getMetadata()` would be better as:

> Metadata which should be applied to the `PodDistruptionBugetTemplate` resource.

(i.e. add `PodDistruptionBugetTemplate` to make it clearer which resource will get the metadata).

scholzj commented:
@tombentley I fixed the API description and removed the table with the second explanation of the `maxUnavailable` field. Please have another look whether you think it is better.

I agree that this to some extent duplicates the API reference. However, I don't think we should just let the users browse the generated reference docs and _discover_ the features from there. I think we need to add at least a minimal mention + link to the reference. So I think having at least a small concept is needed for the user to discover this (I prefer concept over procedure because otherwise we end up with million procedures which from 90% contain the same `oc apply` stuff.).

ppatierno commented:
I think I and @tombentley already had this discussion and my answer was the same as @scholzj 
The auto-generated doc is a reference for the user but while reading the documentation we have to describe stuff and not leaving the user going to the reference doc for understanding things.

tombentley commented:
> So I think having at least a small concept is needed for the user to discover this (I prefer concept over procedure because otherwise we end up with million procedures which from 90% contain the same oc apply stuff.).

I'm fine with having the concept, and I agree it's useful to have somewhere to explain the context of how the properties should be used. I just want to avoid documenting the detail of each property in more than a single place, because it's a direct path to inconsistent documentation.

laidan6000 commented on documentation/book/con-customizing-pod-disruption-budgets.adoc:
```diff
@@ -0,0 +1,33 @@
+
+[id='con-customizing-pod-disruption-budgets-{context}']
+= Customizing Pod Disruption Budgets
+
```
_Please reword the first paragraph as follows:_

AMQ Streams creates a pod disruption budget for every new `StatefulSet` or `Deployment`. By default, these pod disruption budgets only allow a single pod to be unavailable at a given time by setting the `maxUnavailable` value in the`PodDisruptionBudget.spec` resource to 1. You can increase the amount of unavailable pods allowed by using the pod disruption budget template, which applies to each type of cluster (Kafka and Zookeeper; Kafka Connect and Kafka Connect with S2I support; and Kafka Mirror Maker).

scholzj replied:
> You can increase the amount of unavailable pods allowed by changing the default value of maxUnavailable in the pod disruption budget template.

The budget can be increased as well as decreased. So I changed _increased_ to _change_. I hope this is OK for you.

laidan6000 commented on documentation/book/con-customizing-pod-disruption-budgets.adoc:
```diff
@@ -0,0 +1,33 @@
+You can change this default setting using the Pod Disruption Budget template.
+The Pod Disruption Budget template is effective on each type of cluster (Kafka and Zookeeper; Kafka Connect and Kafka Connect with S2I support; and Kafka Mirror Maker).
+
+The following example shows customized Pod Disruption Budget fields on a `template` property:
```
Pod Disruption Budget fields > `podDisruptionBudget` fields

laidan6000 commented on documentation/book/con-customizing-pod-disruption-budgets.adoc:
```diff
@@ -0,0 +1,33 @@
+
+.Additional resources
+
+* For more information, see xref:type-PodDisruptionBudgetTemplate-reference[].
```
* The [Managing Pods](https://access.redhat.com/documentation/en-us/openshift_container_platform/3.6/html/cluster_administration/admin-guide-manage-pods) chapter of the OpenShift Container Platform documentation.

scholzj replied:
As this is upstream documentation, we should use links to Kubernetes. We should change it OpenShift link only in the downstream repo. I used here the link which you asked about in another comment.

laidan6000 commented on documentation/book/con-customizing-template-properties.adoc:
```diff
@@ -48,6 +48,7 @@ When defined in a Kafka cluster, the `template` object can have the following fi
 `perPodService`:: Configures the per-Pod services used by clients connecting to the Kafka broker from outside {ProductPlatformName} to access individual brokers.
 `externalBootstrapRoute`:: Configures the bootstrap route used by clients connecting to the Kafka brokers from outside of {OpenShiftName} using {OpenShiftName} `Routes`.
 `perPodRoute`:: Configures the per-Pod routes used by clients connecting to the Kafka broker from outside {OpenShiftName} to access individual brokers using {OpenShiftName} `Routes`.
+`podDisruptionBudget`:: Configures the Pod Disruption Budget for Kafka broker `StatefulSet`.
```
Lowercase: "Configures the pod disruption budget for..."

laidan6000 commented on documentation/book/con-customizing-template-properties.adoc:
```diff
@@ -57,6 +58,7 @@ When defined in a Zookeeper cluster, the `template` object can have the followin
 `pod`:: Configures the Zookeeper `Pods` created by the `StatefulSet`.
 `clientsService`:: Configures the service used by clients to access Zookeeper.
 `nodesService`:: Configures the headless service.
+`podDisruptionBudget`:: Configures the Pod Disruption Budget for Zookeeper `StatefulSet`.
```
Lowercase: "Configures the pod disruption budget for..."

laidan6000 commented on documentation/book/con-customizing-template-properties.adoc:
```diff
@@ -72,10 +74,12 @@ When used with Kafka Connect and Kafka Connect with Source2Image support , the t
 `deployment`:: Configures the Kafka Connect `Deployment`.
 `pod`:: Configures the Kafka Connect `Pods` created by the `Deployment`.
 `apiService`:: Configures the service used by the Kafka Connect REST API.
+`podDisruptionBudget`:: Configures the Pod Disruption Budget for Kafka Connect `Deployment`.
```
Lowercase: "Configures the pod disruption budget for..."

laidan6000 commented on documentation/book/con-customizing-template-properties.adoc:
```diff
@@ -72,10 +74,12 @@ When used with Kafka Connect and Kafka Connect with Source2Image support , the t
-`pod`:: Configures the Kafka Mirror Maker `Pods` created by the `Deployment`.
\ No newline at end of file
+`pod`:: Configures the Kafka Mirror Maker `Pods` created by the `Deployment`.
+`podDisruptionBudget`:: Configures the Pod Disruption Budget for Kafka Mirror Maker `Deployment`.
```
Lowercase: "Configures the pod disruption budget for..."

laidan6000 commented on documentation/book/ref-list-of-kafka-cluster-resources.adoc:
```diff
@@ -17,12 +17,14 @@ The following resources will created by the Cluster Operator in the {ProductPlat
 `_cluster-name_-kafka-config`:: ConfigMap which contains the Kafka ancillary configuration and is mounted as a volume by the Kafka broker pods.
 `_cluster-name_-kafka-brokers`:: Secret with Kafka broker keys.
 `_cluster-name_-kafka`:: Service account used by the Kafka brokers.
+`_cluster-name_-kafka`:: Pod Disruption Budget configured for the Kafka brokers.
```
Pod disruption budget

laidan6000 commented on documentation/book/ref-list-of-kafka-cluster-resources.adoc:
```diff
@@ -17,12 +17,14 @@ The following resources will created by the Cluster Operator in the {ProductPlat
 `_cluster-name_-zookeeper-client`:: Service used by Kafka brokers to connect to Zookeeper nodes as clients.
 `_cluster-name_-zookeeper-config`:: ConfigMap which contains the Zookeeper ancillary configuration and is mounted as a volume by the Zookeeper node pods.
 `_cluster-name_-zookeeper-nodes`:: Secret with Zookeeper node keys.
+`_cluster-name_-zookeeper`:: Pod Disruption Budget configured for the Zookeeper nodes.
```
Pod disruption budget

laidan6000 commented on documentation/book/ref-list-of-kafka-connect-resources.adoc:
```diff
@@ -10,3 +10,4 @@ The following resources will created by the Cluster Operator in the {ProductPlat
 _connect-cluster-name_-connect:: Deployment which is in charge to create the Kafka Connect worker node pods.
 _connect-cluster-name_-connect-api:: Service which exposes the REST interface for managing the Kafka Connect cluster.
 _connect-cluster-name_-config:: ConfigMap which contains the Kafka Connect ancillary configuration and is mounted as a volume by the Kafka broker pods.
+_connect-cluster-name_-connect:: Pod Disruption Budget configured for the Kafka Connect worker nodes.
```
Pod disruption budget

laidan6000 commented on documentation/book/ref-list-of-kafka-connect-s2i-resources.adoc:
```diff
@@ -13,3 +13,4 @@ _connect-cluster-name_-connect:: ImageStream where the newly built Docker images
 _connect-cluster-name_-connect:: DeploymentConfig which is in charge of creating the Kafka Connect worker node pods.
 _connect-cluster-name_-connect-api:: Service which exposes the REST interface for managing the Kafka Connect cluster.
 _connect-cluster-name_-config:: ConfigMap which contains the Kafka Connect ancillary configuration and is mounted as a volume by the Kafka broker pods.
+_connect-cluster-name_-connect:: Pod Disruption Budget configured for the Kafka Connect worker nodes.
```
Pod disruption budget

laidan6000 commented on documentation/book/ref-list-of-kafka-mirror-maker-resources.adoc:
```diff
@@ -8,4 +8,5 @@
-_<mirror-maker-name>_-config:: ConfigMap which contains the Kafka Mirror Maker ancillary configuration and is mounted as a volume by the Kafka broker pods.
\ No newline at end of file
+_<mirror-maker-name>_-config:: ConfigMap which contains the Kafka Mirror Maker ancillary configuration and is mounted as a volume by the Kafka broker pods.
+_<mirror-maker-name>_-mirror-maker:: Pod Disruption Budget configured for the Kafka Mirror Maker worker nodes.
```
Pod disruption budget

laidan6000 commented on documentation/common/attributes.adoc:
```diff
@@ -53,6 +53,7 @@
 :K8sPullingImagesFromPrivateRegistries: link:https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/[Pull an Image from a Private Registry^]
 :K8sConfigureSecurityContext: link:https://kubernetes.io/docs/tasks/configure-pod-container/security-context/[Configure a Security Context for a Pod or Container^]
 :K8sNetworkPolicyPeerAPI: link:https://v1-9.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#networkpolicypeer-v1-networking[NetworkPolicyPeer API reference^]
+:K8sPodDisruptionBudgets: link:https://kubernetes.io/docs/concepts/workloads/pods/disruptions/[Disruptions^]
```
_Where is this link used?_
